python/ray/autoscaler/v2/autoscaler.py

```diff
@@ -0,0 +1,446 @@
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from ray._raylet import GcsClient
+from ray.autoscaler.v2.instance_manager.config import AutoscalingConfig, IConfigReader
+from ray.autoscaler.v2.instance_manager.instance_manager import (
+    InstanceManager,
+    InstanceUtil,
+)
+from ray.autoscaler.v2.scheduler import (
+    IResourceScheduler,
+    SchedulingReply,
+    SchedulingRequest,
+)
+from ray.core.generated.autoscaler_pb2 import (
+    AutoscalingState,
+    ClusterResourceState,
+    NodeState,
+    NodeStatus,
+)
+from ray.core.generated.instance_manager_pb2 import (
+    GetInstanceManagerStateRequest,
+    Instance,
+    InstanceManagerState,
+    InstanceUpdateEvent,
+    StatusCode,
+    UpdateInstanceManagerStateRequest,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class AutoscalerError(Exception):
+    """Autoscaler error."""
+
+    def __init__(self, msg: str, status: StatusCode) -> None:
+        """
+        Args:
+            msg: The error message.
+            status: The status code.
+        """
+        self.msg = msg
+        self.status = status
+
+
+class IAutoscaler(ABC):
+    """The autoscaler interface."""
+
+    @abstractmethod
+    def get_autoscaling_state(
+        self, cluster_resource_state: ClusterResourceState
+    ) -> Optional[AutoscalingState]:
+        """
+        Get the autoscaling state.
+
+        Args:
+            cluster_resource_state: The cluster resource state that's obtained
+                from GCS. This represents the current state of the ray cluster.
+
+        Returns:
+            The autoscaling state.
+            Or None if the autoscaling state cannot be computed.
+
+        Raises:
+            AutoscalerError: If the autoscaling state cannot be computed.
+        """
+        pass
+
+
+@dataclass
+class AutoscalerInstance:
+    """
+    AutoscalerInstance represents an instance that's managed by the autoscaler.
+    This includes two states:
+        1. the instance manager state: information of the underlying cloud node
+            and the exact status before/after ray runs.
+        2. the ray node state. the ray node like existing resources.
+
+    The two states are linked by the cloud_instance_id.
+
+    It's possible that one state is missing, for example:
+    - Instance state could be missing if the instance manager hasn't discovered
+    the instance yet. (i.e. the node provider didn't inform the instance manager
+    but already started a ray node)
+    - Ray node state could be missing if the ray node has been stopped or hasn't
+    started yet.
+    """
+
+    # The cloud instance id. It could be None if the instance manager hasn't
+    # discovered the instance yet.
+    cloud_instance_id: Optional[str] = None
+    # The instance manager state.
+    im_instance: Optional[Instance] = None
+    # The ray node state.
+    ray_node: Optional[NodeState] = None
```

Comment:
Comment on when ray_node is not None but im_instance is None.

Post init assertion on what None combination is valid.